### PR TITLE
Initial healthchecks implementation

### PR DESCRIPTION
**Pull request by argast**

### CODE REVIEW

File: src/main/scala/com/timeout/healthchecks/HealthChecks.scala [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package com.timeout.healthchecks
+
+import java.net.{Socket, URL}
+
+import cats.syntax.all._
```
Comment (ostapneko) [inline src/main/scala/com/timeout/healthchecks/HealthChecks.scala]: Could we use more targeted imports? It's a good habit going forward so we don't massively increase compile time

File: src/main/scala/com/timeout/healthchecks/HealthChecks.scala [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+import scala.util.Try
+
+object HealthChecks {
+
+  def ping(host: String, port: Int) = healthCheck(s"Ping $host:$port", () => {
```
Comment (ostapneko) [inline src/main/scala/com/timeout/healthchecks/HealthChecks.scala]: Could we also get into the habit of having signature for public functions? It makes it easier to review and again decreases compilation time (I think)

File: src/main/scala/com/timeout/healthchecks/HealthChecks.scala [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    Try {
```
Comment (ostapneko) [inline src/main/scala/com/timeout/healthchecks/HealthChecks.scala]: `ValidatedNel` and `Validated` have these function built-in: `catchOnly` and `catchNonFatal`

File: src/main/scala/com/timeout/healthchecks/package.scala [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import scala.util.Try
+
+package object healthchecks {
+
+  class HealthCheck(val name: String, check: AsyncCheck) {
+
+    val addNameToErrorMessage: String => String = m => s"$name failed: $m"
+
+    def run()(implicit ec: ExecutionContext) = {
+      check().recover {
```
Comment (ostapneko) [inline src/main/scala/com/timeout/healthchecks/package.scala:17]: If `AsyncCheck` already wraps a `ValidatedNel`, is it correct to recover? It looks like either we allow the check to throw and we recover, or it wraps a validation, which means that the error handling should have been done upfront

File: src/main/scala/com/timeout/healthchecks/package.scala [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        case t: Throwable => t.getMessage.invalidNel[Unit]
+      }.map(_.leftMap(r => r.map(addNameToErrorMessage)))
+    }
+  }
+
+  def healthCheck(name: String, c: Check) = new HealthCheck(name, () => Future.fromTry(Try(c())))
```
Comment (ostapneko) [inline src/main/scala/com/timeout/healthchecks/package.scala]: Same as a above, the result shouldn't be double-wrapped

File: src/main/scala/com/timeout/healthchecks/routes/HealthRoutes.scala [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import cats.std.all._
+
+import scala.collection.convert.DecorateAsScala
+import scala.concurrent.{ExecutionContext, Future}
+
+object HealthRoutes extends CirceSupport with DecorateAsScala {
+
+  @JsonCodec case class HealthResult(status: String, checks: Seq[String], failures: Seq[String])
+
+  def status(s: Boolean) = if (s) "OK" else "KO"
```
Comment (ostapneko) [inline src/main/scala/com/timeout/healthchecks/routes/HealthRoutes.scala]: purely for backward compatibility reasons (graffiti convention and zabbix settings), could we have it either "healthy" or "unhealthy" ?

File: src/main/scala/com/timeout/healthchecks/routes/HealthRoutes.scala [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  def statusCode(s: Boolean) = if (s) OK else InternalServerError
+
+  def health(checks: HealthCheck*)(implicit ec: ExecutionContext) = path("health") {
+    complete {
+      Future.traverse(checks.toList)(_.run()).map { r =>
+        val result = r.sequenceU_
+        val healthy = result.isValid
+        val checkNames = checks.map(_.name)
+        val messages: Seq[String] = result.fold(_.map(identity).toList, _ => List())
```
Comment (ostapneko) [inline src/main/scala/com/timeout/healthchecks/routes/HealthRoutes.scala]: nitpick: can this be `errorMessages` so it's a bit clearer?

File: src/test/scala/com/timeout/healthchecks/routes/HealthRoutesTest.scala [hunk 8]
```diff
@@ -0,0 +1,34 @@
+package com.timeout.healthchecks.routes
+
+import akka.http.scaladsl.model.StatusCodes._
+import akka.http.scaladsl.testkit.ScalatestRouteTest
+import org.scalatest.{FunSpec, Matchers, OneInstancePerTest}
+import cats.syntax.all._
+import com.timeout.healthchecks._
+
+class HealthRoutesTest extends FunSpec with ScalatestRouteTest with Matchers with OneInstancePerTest {
```
Comment (ostapneko) [inline src/test/scala/com/timeout/healthchecks/routes/HealthRoutesTest.scala:9]: 👍

File: src/main/scala/com/timeout/healthchecks/HealthChecks.scala (hunk 1 above)
Comment (argast) [inline src/main/scala/com/timeout/healthchecks/HealthChecks.scala]: Fixed

File: src/main/scala/com/timeout/healthchecks/routes/HealthRoutes.scala (hunk 6 above)
Comment (argast) [inline src/main/scala/com/timeout/healthchecks/routes/HealthRoutes.scala]: done

File: src/main/scala/com/timeout/healthchecks/routes/HealthRoutes.scala (hunk 7 above)
Comment (argast) [inline src/main/scala/com/timeout/healthchecks/routes/HealthRoutes.scala]: done

File: src/main/scala/com/timeout/healthchecks/HealthChecks.scala (hunk 2 above)
Comment (argast) [inline src/main/scala/com/timeout/healthchecks/HealthChecks.scala]: sure, fixed.

File: src/main/scala/com/timeout/healthchecks/HealthChecks.scala (hunk 3 above)
Comment (argast) [inline src/main/scala/com/timeout/healthchecks/HealthChecks.scala]: ValidatedNel is just a simple type alias, but I managed to use Validated.catchNonFatal and wrap result in NonEmptyList.

**Pull request merged**